packages/world/ts/objects.ts

```diff
@@ -0,0 +1,520 @@
+export const numBlockCategories = 128 / 2;
+
+export const blockCategories = [
+  // terrain categories
+  "NonSolid",
+  "Stone",
+  "Gemstone",
+  "Soil",
+  "Ore",
+  "Sand",
+  "Clay",
+  "Log",
+  "Leaf",
+  "Flower",
+  "Greenery",
+  "Crop",
+  "CropBlock",
+  "UnderwaterPlant",
+  "UnderwaterBlock",
+  // non-terrain categories
+  "Plank",
+  "OreBlock",
+  "Seed",
+  "Sapling",
+  "Station",
+  "SmartEntityBlock",
+  "MiscBlock",
+] as const;
+
+export const nonBlockCategories = [
+  "Pick",
+  "Axe",
+  "Hoe",
+  "Whacker",
+  "OreBar",
+  "Bucket",
+  "Food",
+  "Fuel",
+  "Movable",
+  "SmartEntityNonBlock",
+] as const;
+
+export const allCategories = [
+  ...blockCategories,
+  ...nonBlockCategories,
+] as const;
+
+export type Category = (typeof allCategories)[number];
+
+// Define categories that pass metadata for the template
+export interface CategoryMetadata {
+  name: string;
+  index: number;
+}
+
+// Transform block categories into metadata objects
+export const blockCategoryMetadata: CategoryMetadata[] = blockCategories.map(
+  (name, index) => ({
+    name,
+    index,
+  }),
+);
+
+// Transform non-block categories into metadata objects
+export const nonBlockCategoryMetadata: CategoryMetadata[] =
+  nonBlockCategories.map((name, index) => ({
+    name,
+    index: numBlockCategories + index, // Start after block categories
+  }));
+
+// All categories metadata for template generation
+export const allCategoryMetadata: CategoryMetadata[] = [
+  ...blockCategoryMetadata,
+  ...nonBlockCategoryMetadata,
+];
+
+// Meta-categories (categories that should be included in pass-through check)
+export const passThroughCategories: Category[] = [
+  "NonSolid",
+  "Leaf",
```

Comment:
All the leaves, bamboo_bush, and cactus are not pass through

Reply:
ah true true